{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### For perfect"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "figsuplix = \"pdf\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def is_Exist_file(path):\n",
    "    import os\n",
    "    if os.path.exists(path):\n",
    "        os.remove(path)\n",
    "\n",
    "\n",
    "def mkdir(path):\n",
    "    import os\n",
    "    path = path.strip()  # 去除首位空格\n",
    "    path = path.rstrip(\"\\\\\")  # 去除尾部 \\ 符号\n",
    "    isExists = os.path.exists(path)  # 判断路径是否存在\n",
    "    # 判断结果\n",
    "    if not isExists:\n",
    "        os.makedirs(path)  # 如果不存在则创建目录\n",
    "        print(path + ' 创建成功')\n",
    "    else:\n",
    "        print(path + ' 目录已存在')  # 如果目录存在则不创建，并提示目录已存在"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "## mutation 坐标转换\n",
    "#####################################################################################\n",
    "## 反向互补\n",
    "def reverseComplement(seq):\n",
    "    \"\"\"\n",
    "     生成反向互补序列\n",
    "     :param seq:\n",
    "     :return:revComSeq\n",
    "     \"\"\"\n",
    "    ATCG_dict = {'A': 'T', 'T': 'A', 'C': 'G', 'G': 'C', 'a': 't', 't': 'a', 'c': 'g', 'g': 'c', 'N': 'N'}\n",
    "    revComSeq = ''\n",
    "    for i in seq:\n",
    "        revComSeq = ATCG_dict[i] + revComSeq\n",
    "    return revComSeq\n",
    "\n",
    "\n",
    "## 转换核心函数\n",
    "def transformer_core(mut, index_dict):\n",
    "    if 'M' in mut:\n",
    "        inf, sup = int(mut.split(':')[0][1:]), int(mut.split(':')[1][:3])\n",
    "        bn, an = mut.split('_')[1].split(':')[0], mut.split('_')[1].split(':')[1]\n",
    "        new_inf, new_sup = index_dict[inf], index_dict[sup]\n",
    "        new_bn, new_an = reverseComplement(bn), reverseComplement(an)\n",
    "        return '%s%s:%sM_%s:%s'%(mut[0], new_inf, new_sup, new_bn, new_an)\n",
    "    elif 'I' in mut:\n",
    "        new_inser_site = index_dict[int(mut.split('_')[0][1:-1])]\n",
    "        new_inser_nucles = reverseComplement(mut.split('_')[-1])\n",
    "        return '%s%sI_%s'%(mut[0], new_inser_site, new_inser_nucles)\n",
    "    elif 'D' in mut:\n",
    "        inf, sup = int(mut.split(':')[0][1:]), int(mut.split(':')[-1][:-1])\n",
    "        new_inf, new_sup = index_dict[inf], index_dict[sup]\n",
    "        return '%s%s:%sD'%(mut[0], new_inf, new_sup)\n",
    "    else:\n",
    "        return mut\n",
    "\n",
    "\n",
    "## 解析 all_mutation & target_mutation 信息\n",
    "def parse_mutation_information(mutation):\n",
    "    '''\n",
    "    mutation = 'X0+Y179:180D.Y186:186M_C:A+Z196I_T'\n",
    "    # mutation = 'X0+Y0+Z0'\n",
    "    muts = parse_mutation_information(mutation)\n",
    "    '''\n",
    "    import re\n",
    "    mut_list = re.split('\\+|\\.', mutation)\n",
    "    mut_dict = {'X': [], 'Y': [], 'Z': []}\n",
    "    for mut in mut_list:\n",
    "        mut_dict[mut[0]].append(mut)\n",
    "    return mut_dict\n",
    "\n",
    "\n",
    "## 调整突变坐标的参考系：以 gRNASeq 为基\n",
    "def adjust_off_target_mutaion_reference_system(mutation):\n",
    "    '''\n",
    "    mutation = 'X164I_T+Y178I_C+Z0'\n",
    "    mutation = 'X164:164D+Y187I_G+Z199:199M_C:T'\n",
    "    mutation = 'X0+Y0+Z0'\n",
    "    adjust_off_target_mutaion_reference_system(mutation)\n",
    "    '''\n",
    "    ## 长序列与短序列对应字典\n",
    "    index_dict = {}\n",
    "    for i in range(20):\n",
    "        index_dict[125 + i] = -19 + i\n",
    "    for i in range(63):\n",
    "        index_dict[145 + i] = 1 + i\n",
    "    for i in range(20):\n",
    "        index_dict[208 + i] = 64 + i\n",
    "    ## 解析 all_mutation & target_mutation 信息\n",
    "    mut_dict = parse_mutation_information(mutation)\n",
    "    new_mut_dict = {'X': '', 'Y': '', 'Z': ''}\n",
    "    for label, mut_list in mut_dict.items():\n",
    "        a_list = []\n",
    "        for mut in mut_list:\n",
    "            new_mut = transformer_core(mut, index_dict)\n",
    "            a_list.append(new_mut)\n",
    "        a_list.sort(reverse=False)\n",
    "        new_mut_dict[label] = '.'.join(a_list)\n",
    "    new_list = []\n",
    "    for label in ['X', 'Y', 'Z']:\n",
    "        new_list.append(new_mut_dict[label])\n",
    "    return '+'.join(new_list)\n",
    "#####################################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "## up, core, down 区域内突变核苷酸总长度\n",
    "def region_mutation_length(mutation):\n",
    "    mut_list = mutation.split(\".\")\n",
    "    mut_length = 0\n",
    "    for mut in mut_list:\n",
    "        if 'M' in mut:\n",
    "            mut_length += len(mut.split(\":\")[-1])\n",
    "        elif 'I' in mut:\n",
    "            mut_length += len(mut.split(\"_\")[-1])\n",
    "        elif 'D' in mut:\n",
    "            inf, sup = int(mut.split(':')[0][1:]), int(mut.split(':')[-1][:-1])\n",
    "            delt_length = sup - inf + 1\n",
    "            mut_length += delt_length\n",
    "        else:\n",
    "            pass\n",
    "    return mut_length\n",
    "\n",
    "\n",
    "def mutation_region_splitting(data):\n",
    "    ## step 1: mutaion 区域划分\n",
    "    data['new_mutation'] = data['all_mutation'].apply(lambda x: adjust_off_target_mutaion_reference_system(x))\n",
    "    ## up, core, down 区域内突变类型\n",
    "    data['up_mutation'] = data['new_mutation'].apply(lambda x: x.split('+')[0])\n",
    "    data['core_mutation'] = data['new_mutation'].apply(lambda x: x.split('+')[1])\n",
    "    data['down_mutation'] = data['new_mutation'].apply(lambda x: x.split('+')[2])\n",
    "    ## up, core, down 区域内间断突变的个数\n",
    "    data['up_mut_num'] = data['up_mutation'].apply(lambda x: len(x.split('.')) if x != 'X0' else 0)\n",
    "    data['core_mut_num'] = data['core_mutation'].apply(lambda x: len(x.split('.')) if x != 'Y0' else 0)\n",
    "    data['down_mut_num'] = data['down_mutation'].apply(lambda x: len(x.split('.')) if x != 'Z0' else 0)\n",
    "    ## up, core, down 区域内突变核苷酸总长度\n",
    "    data['up_mut_length'] = data['up_mutation'].apply(lambda x: region_mutation_length(x))\n",
    "    data['core_mut_length'] = data['core_mutation'].apply(lambda x: region_mutation_length(x))\n",
    "    data['down_mut_length'] = data['down_mutation'].apply(lambda x: region_mutation_length(x))\n",
    "    ############################################################\n",
    "    print(\"\\nFirst: 统计 X0+Y0+Z0、XA+Y0+Z0、X0+YA+Z0、X0+Y0+ZA、XA+YA+Z0、XA+Y0+ZA、X0+YA+ZA、XA+YA+ZA 的类型占比\")\n",
    "    data_000 = data.loc[(data['up_mut_num']==0) & (data['core_mut_num']==0) & (data['down_mut_num']==0), :]\n",
    "    data_100 = data.loc[(data['up_mut_num']!=0) & (data['core_mut_num']==0) & (data['down_mut_num']==0), :]\n",
    "    data_010 = data.loc[(data['up_mut_num']==0) & (data['core_mut_num']!=0) & (data['down_mut_num']==0), :]\n",
    "    data_001 = data.loc[(data['up_mut_num']==0) & (data['core_mut_num']==0) & (data['down_mut_num']!=0), :]\n",
    "    data_110 = data.loc[(data['up_mut_num']!=0) & (data['core_mut_num']!=0) & (data['down_mut_num']==0), :]\n",
    "    data_101 = data.loc[(data['up_mut_num']!=0) & (data['core_mut_num']==0) & (data['down_mut_num']!=0), :]\n",
    "    data_011 = data.loc[(data['up_mut_num']==0) & (data['core_mut_num']!=0) & (data['down_mut_num']!=0), :]\n",
    "    data_111 = data.loc[(data['up_mut_num']!=0) & (data['core_mut_num']!=0) & (data['down_mut_num']!=0), :]\n",
    "    ## count\n",
    "    count_dict = {}\n",
    "    count_000 = data_000.shape[0]\n",
    "    count_100, count_010, count_001 = data_100.shape[0], data_010.shape[0], data_001.shape[0]\n",
    "    count_110, count_101, count_011 = data_110.shape[0], data_101.shape[0], data_011.shape[0]\n",
    "    count_111 = data_111.shape[0]\n",
    "    count_dict['X0+Y0+Z0'] = count_000\n",
    "    count_dict['XA+Y0+Z0'] = count_100\n",
    "    count_dict['X0+YA+Z0'] = count_010\n",
    "    count_dict['X0+Y0+ZA'] = count_001\n",
    "    count_dict['XA+YA+Z0'] = count_110\n",
    "    count_dict['XA+Y0+ZA'] = count_101\n",
    "    count_dict['X0+YA+ZA'] = count_011\n",
    "    count_dict['XA+YA+ZA'] = count_111\n",
    "    print(\"X0+Y0+Z0:\", count_000)\n",
    "    print(\"XA+Y0+Z0:\", count_100)\n",
    "    print(\"X0+YA+Z0:\", count_010)\n",
    "    print(\"X0+Y0+ZA:\", count_001)\n",
    "    print(\"XA+YA+Z0:\", count_110)\n",
    "    print(\"XA+Y0+ZA:\", count_101)\n",
    "    print(\"X0+YA+ZA:\", count_011)\n",
    "    print(\"XA+YA+ZA:\", count_111)\n",
    "    count_sum = data.shape[0]\n",
    "    print(\"Distribution of count ratio:\", round(count_000/count_sum, 3), \n",
    "                        round(count_100/count_sum, 3), round(count_010/count_sum, 3), round(count_001/count_sum, 3), \n",
    "                        round(count_110/count_sum, 3), round(count_101/count_sum, 3), round(count_011/count_sum, 3), \n",
    "                        round(count_111/count_sum, 3))\n",
    "    return (data_000, data_100, data_010, data_001, data_110, data_101, data_011, data_111, count_dict)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 1.1 on-target 编辑效率分析\n",
    "def distplot_on_target_efficiency(data_000, cell_line, replicate, save_dir):\n",
    "    import seaborn as sns\n",
    "    import matplotlib.pyplot as plt\n",
    "    sns.set(color_codes=True)\n",
    "    fig, ax = plt.subplots(1,1, figsize=(6, 5))\n",
    "    # 为坐标轴设置名称\n",
    "    colors = ['slategrey','lightsteelblue','cornflowerblue','royalblue','darkorange','lightgreen',\n",
    "         'green','r','teal','blue','slateblue','darkslateblue',\n",
    "         'maroon','blueviolet','indigo','darkorchid','mediumorchid','thistle',\n",
    "         'plum','violet','purple','m','lightseagreen','magenta',\n",
    "         'orchid','chartreuse','deeppink','hotpink','palevioletred','pink']\n",
    "    sns.distplot(data_000['off-target_eff'], hist=False, color='purple')\n",
    "    plt.xlabel(\"on-target efficiency\", fontsize=12, weight='bold')\n",
    "    plt.ylabel(\"Density\",fontsize=12, weight='bold')\n",
    "    title = \"%s-%s on-target editing efficiency\"%(cell_line, replicate)\n",
    "    plt.title(title, fontsize=12, weight='bold')\n",
    "    savefig_path = '%s/%s.%s'%(save_dir, title, figsuplix)\n",
    "    plt.savefig(savefig_path, dpi=300, bbox_inches = 'tight')\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### For mismatch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "########################################################\n",
    "def is_Exist_file(path):\n",
    "    import os\n",
    "    if os.path.exists(path):\n",
    "        os.remove(path)\n",
    "\n",
    "\n",
    "def mkdir(path):\n",
    "    import os\n",
    "    path = path.strip()  # 去除首位空格\n",
    "    path = path.rstrip(\"\\\\\")  # 去除尾部 \\ 符号\n",
    "    isExists = os.path.exists(path)  # 判断路径是否存在\n",
    "    # 判断结果\n",
    "    if not isExists:\n",
    "        os.makedirs(path)  # 如果不存在则创建目录\n",
    "        print(path + ' 创建成功')\n",
    "    else:\n",
    "        print(path + ' 目录已存在') "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 1. distribution of off-target mismatch"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### 1.1 distribution of the mismatch number"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "## core 区域内 mutation 的长度\n",
    "def get_off_target_mismatch_number(data_010):\n",
    "    core_mut_length_dict = {}\n",
    "    mut_length_list = list(data_010['core_mut_length'].unique())\n",
    "    for mut_length in mut_length_list: \n",
    "        templ = data_010.loc[data_010['core_mut_length']==mut_length, :].shape[0]\n",
    "        core_mut_length_dict['core_mut_length_' + str(mut_length)] = round(templ/data_010.shape[0], 3)\n",
    "    print(\"\\ncore mutation length:\")\n",
    "    print(core_mut_length_dict)\n",
    "    return core_mut_length_dict\n",
    "\n",
    "\n",
    "## plot distribution of off-target mismatch number\n",
    "def plot_distribution_off_target_mismacth_num(core_mut_length_dict, save_dir):\n",
    "    data_list = []\n",
    "    labels = []\n",
    "    keys = list(core_mut_length_dict.keys())\n",
    "    keys.sort(reverse=False)\n",
    "    for k in keys:\n",
    "        v = core_mut_length_dict[k]\n",
    "        data_list.append(round(v*100, 2))\n",
    "        labels.append(k.split('_')[-1])\n",
    "\n",
    "    ## 柱状图\n",
    "    import matplotlib.pyplot as plt\n",
    "    # 设置默认绘图风格\n",
    "    plt.style.use(\"seaborn-white\")  \n",
    "    fig, ax = plt.subplots(1,1, figsize=(6, 4))\n",
    "\n",
    "    plt.bar(range(len(data_list)), data_list)\n",
    "    ## 坐标轴不可见\n",
    "    ax.spines['top'].set_visible(False)\n",
    "    ax.spines['right'].set_visible(False)\n",
    "    ## xlabel, ylabel\n",
    "    plt.xlabel(\"Mismatch length of Target+PAM Region\", fontsize=12, weight='bold')\n",
    "    plt.ylabel(\"Ratio (%)\", fontsize=12, weight='bold')\n",
    "    ## xticks\n",
    "    plt.xticks(range(len(labels)), labels, fontsize=12, weight='bold')\n",
    "    ## text\n",
    "    ## title\n",
    "    title = 'distribution of off-target mismatch number'\n",
    "    plt.title(title, fontsize=12, weight='bold')\n",
    "    ## save\n",
    "    mkdir(save_dir)\n",
    "    savefig_path = save_dir + '/%s.%s'%(title, figsuplix)\n",
    "    plt.savefig(savefig_path, dpi=300, bbox_inches = 'tight')\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### 1.2 distribution of off-target mismatch among Target + PAM regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 确定 mutation site 的上下界限\n",
    "def mutation_site_range(mut_site):\n",
    "    if 'M' in mut_site:\n",
    "        mutk = mut_site.split(\"_\")[0].split(':')\n",
    "        inf, sup = int(mutk[0][1:]), int(mutk[1][:-1])\n",
    "    elif 'I' in mut_site:\n",
    "        mut = mut_site.split(\"_\")[0]\n",
    "        inf, sup = int(mut[1:-1]), int(mut[1:-1])\n",
    "    elif 'D' in mut_site: ## deletion\n",
    "        mut = mut_site.split(\":\")\n",
    "        inf, sup = int(mut[0][1:]), int(mut[1][:-1])\n",
    "    else: ## X0/Y0/Z0\n",
    "        inf, sup = 0, 0\n",
    "    return (inf, sup)\n",
    "\n",
    "\n",
    "## 计算 core 区域 mut_num == x 的位置分布\n",
    "def mutation_length_x_position_distribution(data):\n",
    "    pos_dict = {}\n",
    "    for i in range(0, 63, 1):\n",
    "        pos_dict[i] = 0\n",
    "    for index, row in data.iterrows():\n",
    "        mut = row['core_mutation']\n",
    "        inf, sup = mutation_site_range(mut)\n",
    "        for i in range(inf, sup + 1, 1):\n",
    "            pos_dict[i] += 1\n",
    "    sum_count = sum(list(pos_dict.values()))\n",
    "    pos_ratio_dict = {}\n",
    "    for pos, count in pos_dict.items():\n",
    "        pos_ratio_dict[pos] = count/sum_count\n",
    "    return pos_ratio_dict\n",
    "\n",
    "\n",
    "## plot distribution of off-target mismatch among Target + PAM regions\n",
    "def plot_distribution_off_target_mismatch_with_positions(data, save_dir):\n",
    "    for x in range(1, 7, 1):\n",
    "        data_x = data.loc[data['core_mut_length']==x, :]\n",
    "        data_x.reset_index(drop=True, inplace=True)\n",
    "        pos_ratio_dict = mutation_length_x_position_distribution(data_x)\n",
    "        print(\"YA 内的 mismatch number == x  的位置分布:\")\n",
    "        # print(pos_ratio_dict)\n",
    "        ## 柱状图展示 mutation 的位置分布\n",
    "        labels = []\n",
    "        data_list = []\n",
    "        for k in range(21, 44):\n",
    "            if k == 41:\n",
    "                labels.append(\"N\")\n",
    "            elif (k == 42) | (k == 43):\n",
    "                labels.append(\"G\")\n",
    "            else:\n",
    "                labels.append(k-20)\n",
    "            data_list.append(pos_ratio_dict[k])\n",
    "        ############################################################\n",
    "        ## 柱状图\n",
    "        import matplotlib.pyplot as plt\n",
    "        # 设置默认绘图风格\n",
    "        plt.style.use(\"seaborn-white\")  \n",
    "        fig, ax = plt.subplots(1,1, figsize=(12, 4))\n",
    "\n",
    "        plt.bar(range(len(data_list)), data_list, color='lightseagreen')\n",
    "        ## 坐标轴不可见\n",
    "        ax.spines['top'].set_visible(False)\n",
    "        ax.spines['right'].set_visible(False)\n",
    "        ## xlabel, ylabel\n",
    "        plt.xlabel(\"Target+PAM Region\", fontsize=12, weight='bold')\n",
    "        plt.ylabel(\"Ratio\", fontsize=12, weight='bold')\n",
    "        ## xticks\n",
    "        plt.xticks(range(len(labels)), labels, fontsize=12, weight='bold')\n",
    "        ## text\n",
    "        ## title\n",
    "        title = 'position distribution of off-target mismatch_number-%s'%(x)\n",
    "        plt.title(title, fontsize=12, weight='bold')\n",
    "        ## save\n",
    "        mkdir(save_dir)\n",
    "        savefig_path = save_dir + '/%s.%s'%(title, figsuplix)\n",
    "        plt.savefig(savefig_path, dpi=300, bbox_inches = 'tight')\n",
    "        plt.show()\n",
    "############################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### 1.3. distribution of substitution types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "## step 2.3: substitution 的类型\n",
    "## core region 内的核苷酸替换类型\n",
    "def obtain_substitution_type(data):\n",
    "    sub_dict = {'A->C': 0, 'A->G': 0, 'A->T': 0, \n",
    "                'C->A': 0, 'C->G': 0, 'C->T': 0, \n",
    "                'G->A': 0, 'G->C': 0, 'G->T': 0, \n",
    "                'T->A': 0, 'T->C': 0, 'T->G': 0}\n",
    "    for index, row in data.iterrows():\n",
    "        mutation = row['core_mutation']\n",
    "        import re\n",
    "        mut_list = re.split('\\+|\\.', mutation)\n",
    "        for mut in mut_list:\n",
    "            if 'M' in mut:\n",
    "                subs = mut.split('_')[-1].split(':')\n",
    "                ori_n, src_n = subs[0], subs[1]\n",
    "                for index, n in enumerate(ori_n):\n",
    "                    sub = '%s->%s'%(n, src_n[index])\n",
    "                    sub_dict[sub] += 1\n",
    "            else:\n",
    "                pass\n",
    "    ## ratio\n",
    "    sum_c = sum(list(sub_dict.values()))\n",
    "    new_sub_dict = {}\n",
    "    for k, v in sub_dict.items():\n",
    "        new_sub_dict[k] = v/sum_c\n",
    "    return new_sub_dict\n",
    "\n",
    "\n",
    "####################################################################\n",
    "## mismatch length == x\n",
    "def plot_distribution_substitution_types(data, save_dir):\n",
    "    for x in range(1, 7, 1): \n",
    "        data_x = data.loc[data['core_mut_length']==x, :]\n",
    "        data_x.reset_index(drop=True, inplace=True)\n",
    "        sub_dict = obtain_substitution_type(data_x)\n",
    "        print('core region 内的核苷酸替换类型:')\n",
    "        # print(sub_dict)\n",
    "        labels = []\n",
    "        data_list = []\n",
    "        for label, v in sub_dict.items():\n",
    "            labels.append(label)\n",
    "            data_list.append(v)\n",
    "\n",
    "        ####################################################################\n",
    "        ## 柱状图\n",
    "        import matplotlib.pyplot as plt\n",
    "        # 设置默认绘图风格\n",
    "        plt.style.use(\"seaborn-white\")  \n",
    "        fig, ax = plt.subplots(1,1, figsize=(12, 4))\n",
    "\n",
    "        plt.bar(range(len(data_list)), data_list, color='slateblue')\n",
    "        ## 坐标轴不可见\n",
    "        ax.spines['top'].set_visible(False)\n",
    "        ax.spines['right'].set_visible(False)\n",
    "        ## xlabel, ylabel\n",
    "        plt.xlabel(\"Target+PAM Region\", fontsize=12, weight='bold')\n",
    "        plt.ylabel(\"Ratio\", fontsize=12, weight='bold')\n",
    "        ## xticks\n",
    "        plt.xticks(range(len(labels)), labels, fontsize=12, weight='bold')\n",
    "        ## text\n",
    "        ## title\n",
    "        title = 'Distribution of nucleotide substitution type of off-target mismatch_number-%s'%(x)\n",
    "        plt.title(title, fontsize=12, weight='bold')\n",
    "        ## save\n",
    "        mkdir(save_dir)\n",
    "        savefig_path = save_dir + '/%s.%s'%(title, figsuplix)\n",
    "        plt.savefig(savefig_path, dpi=300, bbox_inches = 'tight')\n",
    "        plt.show()\n",
    "####################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 2. Heatmap of off-target single mismatch analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "## mismatch length == x\n",
    "## 确定 mutation site 的上下界限\n",
    "def mutation_site_range(mut_site):\n",
    "    if 'M' in mut_site:\n",
    "        mutk = mut_site.split(\"_\")[0].split(':')\n",
    "        inf, sup = int(mutk[0][1:]), int(mutk[1][:-1])\n",
    "    elif 'I' in mut_site:\n",
    "        mut = mut_site.split(\"_\")[0]\n",
    "        inf, sup = int(mut[1:-1]), int(mut[1:-1])\n",
    "    else: ## deletion\n",
    "        mut = mut_site.split(\":\")\n",
    "        inf, sup = int(mut[0][1:]), int(mut[1][:-1])\n",
    "    return (inf, sup)\n",
    "\n",
    "\n",
    "## get mismatch position \n",
    "def get_off_target_position(mutation):\n",
    "    import re\n",
    "    mut_list = re.split('\\+|\\.', mutation)\n",
    "    mut_list.sort(reverse=False)\n",
    "    position_t = []\n",
    "    for mut_site in mut_list:\n",
    "        inf, sup = mutation_site_range(mut_site)\n",
    "        for i in range(inf, sup + 1):\n",
    "            position_t.append(i)\n",
    "    return tuple(position_t)\n",
    "\n",
    "\n",
    "## get mismatch substitution type\n",
    "def get_mismatch_substitution(mutation):\n",
    "    import re\n",
    "    mut_list = re.split('\\+|\\.', mutation)\n",
    "    mut_list.sort(reverse=False)\n",
    "    subs = []\n",
    "    for mut in mut_list:\n",
    "        if 'M' in mut:\n",
    "            ori_nucles, des_nucles = mut.split('_')[-1].split(':')[0], mut.split('_')[-1].split(':')[1]\n",
    "            for i, nucle in enumerate(ori_nucles):\n",
    "                sub = (nucle, des_nucles[i])\n",
    "                subs.append(sub)\n",
    "    return tuple(subs)\n",
    "\n",
    "\n",
    "## 获得 mismatch length == 1 的某一位置的 heatmap data\n",
    "def get_plot_mismatch1_heatmap_data(pos, data_010, on_data):\n",
    "    ## parse position_tuple & substitution \n",
    "    data_010_x = data_010.loc[data_010['core_mut_length'] == 1, :]\n",
    "    data_010_x['position_tuple'] = data_010_x['core_mutation'].apply(lambda x: get_off_target_position(x))\n",
    "    data_010_x['substitution'] = data_010_x['core_mutation'].apply(lambda x: get_mismatch_substitution(x))\n",
    "    ## merge off-data & on-data\n",
    "    on_data.rename(columns={'off-target_eff': 'on-target_eff'}, inplace=True)\n",
    "    data_010_x = pd.merge(data_010_x, on_data[['sgRNA_name', 'on-target_eff']], how='inner', on='sgRNA_name')\n",
    "    data_010_x['reltv_eff'] = data_010_x.apply(lambda row: row['off-target_eff']/row['on-target_eff'], axis=1)\n",
    "    ## get plot data\n",
    "    position_tuple = (pos,)\n",
    "    pdata = data_010_x.loc[data_010_x['position_tuple'] == position_tuple, :]\n",
    "    subs_list = list(pdata['substitution'].unique()) ## 所有可能的替换类型\n",
    "    stat_count = pd.DataFrame({'A': [0]*4, 'C': [0]*4, 'G': [0]*4, 'T': [0]*4})\n",
    "    plot_data = pd.DataFrame({'A': [0]*4, 'C': [0]*4, 'G': [0]*4, 'T': [0]*4})\n",
    "    plot_data.index = ['A', 'C', 'G', 'T']\n",
    "    stat_count.index = ['A', 'C', 'G', 'T']\n",
    "    for subs in subs_list:\n",
    "        for index, sub in enumerate(subs):\n",
    "            position = position_tuple[index]\n",
    "            ori_nucle, des_nucle = sub ## 列是原始核苷酸，行是替换的核苷酸\n",
    "            plot_data.loc[des_nucle, ori_nucle] = pdata.loc[pdata['substitution'] == subs, 'reltv_eff'].median()\n",
    "            stat_count.loc[des_nucle, ori_nucle] = pdata.loc[pdata['substitution'] == subs, :].shape[0]/data_010_x.shape[0]\n",
    "    return plot_data, stat_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 确定所以值得最大、最小\n",
    "def set_vmin_max(data_010, on_data, data_id, vmin, vmax, ori_pos=21, des_pos=44):\n",
    "    for pos in range(ori_pos, des_pos, 1):\n",
    "        data_tuple = get_plot_mismatch1_heatmap_data(pos, data_010, on_data)\n",
    "        data = data_tuple[data_id]\n",
    "        vmin = min([np.min(np.array(data)), vmin])\n",
    "        vmax = max([np.max(np.array(data)), vmax])\n",
    "    return vmin, vmax\n",
    "\n",
    "\n",
    "## plot heatmap\n",
    "def help_creak_mask(plot_data):\n",
    "    mask = np.zeros_like(plot_data)\n",
    "    for i in range(plot_data.shape[0]):\n",
    "        for j in range(plot_data.shape[1]):\n",
    "            # print(i, j, plot_data.iloc[i, j])\n",
    "            if plot_data.iloc[i, j] == 0:\n",
    "                mask[i, j] = 0\n",
    "    return mask\n",
    "\n",
    "\n",
    "## 展示 mismatch 的 position-substitution 的分布\n",
    "## For position 1 to 21\n",
    "def plot_multiheatmap_ratio_1(data_010, on_data, cell_line, replicate, suffix, title, save_dir, mask, \n",
    "                            vmin=0, vmax=1, data_id=0, cmap=\"Greys\"):\n",
    "    ## position annotation: transforme\n",
    "    tf_pos_dict = {}\n",
    "    for i in range(21, 44, 1):\n",
    "        if i == 41:\n",
    "            tf_pos_dict[i] = \"N\"\n",
    "        elif (i == 42) | (i == 43):\n",
    "            tf_pos_dict[i] = \"G\"\n",
    "        else:\n",
    "            tf_pos_dict[i] = i-20\n",
    "    \n",
    "    import matplotlib.pyplot as plt\n",
    "    import matplotlib.gridspec as gridspec\n",
    "    import seaborn as sns\n",
    "    sns.set_style(\"white\")\n",
    "    sns.set(font_scale=0.6) ## 修改 seaborn 字体比例\n",
    "    fig = plt.figure(figsize=(15, 1.2))\n",
    "    k = 23\n",
    "    gs = gridspec.GridSpec(2, k, height_ratios=[1, 1]) ## 多子图构建，2行 k列\n",
    "\n",
    "    ## 调整子图间的整体布局\n",
    "    fig.tight_layout() # 调整整体空白\n",
    "    plt.subplots_adjust(wspace=0.01, hspace=0) # 调整子图间距\n",
    "\n",
    "    ytick=['A', 'C', 'G', 'T']\n",
    "    ## start plot: the left\n",
    "    pos = 21\n",
    "    data_tuple = get_plot_mismatch1_heatmap_data(pos, data_010, on_data)\n",
    "    data = data_tuple[data_id]\n",
    "    ax0 = plt.subplot(gs[0]) ## 确定子图位置\n",
    "    sns.heatmap(ax=ax0, data=data, mask=mask, vmin=vmin, vmax=vmax, square=True, cbar=False, yticklabels=ytick, cmap=cmap)\n",
    "    ax0.set_xlabel(tf_pos_dict[pos], fontsize=12, weight='bold')\n",
    "    for i in range(1, k-3):\n",
    "        pos += 1\n",
    "        data_tuple = get_plot_mismatch1_heatmap_data(pos, data_010, on_data)\n",
    "        data = data_tuple[data_id]\n",
    "        ax1 = plt.subplot(gs[i]) ## 确定子图位置\n",
    "        sns.heatmap(ax=ax1, data=data, mask=mask, vmin=vmin, vmax=vmax, square=True, cbar=False, yticklabels=[], cmap=cmap)\n",
    "        ax1.set_xlabel(tf_pos_dict[pos], fontsize=10, weight='bold')\n",
    "        if pos == 32:\n",
    "            plt.title(title, fontsize=10, weight='bold')\n",
    "    ## end plot: the right N\n",
    "    sns.set(font_scale=0.5) ## 修改 seaborn 字体比例\n",
    "    pos += 1\n",
    "    data_tuple = get_plot_mismatch1_heatmap_data(pos, data_010, on_data)\n",
    "    data = data_tuple[data_id]\n",
    "    ax1 = plt.subplot(gs[k-3]) ## 确定子图位置\n",
    "    sns.heatmap(ax=ax1, data=data, mask=mask, vmin=vmin, vmax=vmax, square=False, \n",
    "                cbar=True, cbar_kws={\"shrink\": 1, 'ticks': [0.,0.04,0.08,0.12]}, \n",
    "                yticklabels=[], cmap=cmap)\n",
    "    ax1.set_xlabel(tf_pos_dict[pos], fontsize=10, weight='bold')\n",
    "    ## title\n",
    "    # plt.title(title, fontsize=10, weight='bold')\n",
    "    mkdir(save_dir)\n",
    "    savefig_path = save_dir + '/%s.%s'%(title, figsuplix)\n",
    "    plt.savefig(savefig_path, dpi=300, bbox_inches = 'tight')\n",
    "\n",
    "    \n",
    "\n",
    "## 展示 mismatch 的 position-substitution 的分布\n",
    "## For PAM-GG\n",
    "def plot_multiheatmap_ratio_2(data_010, on_data, cell_line, replicate, suffix, title, save_dir, mask, \n",
    "                            vmin=0, vmax=1, data_id=0, cmap=\"Greys\"):\n",
    "    ## position annotation: transforme\n",
    "    tf_pos_dict = {}\n",
    "    for i in range(21, 44, 1):\n",
    "        if i == 41:\n",
    "            tf_pos_dict[i] = \"N\"\n",
    "        elif (i == 42) | (i == 43):\n",
    "            tf_pos_dict[i] = \"G\"\n",
    "        else:\n",
    "            tf_pos_dict[i] = i-20\n",
    "    \n",
    "    import matplotlib.pyplot as plt\n",
    "    import matplotlib.gridspec as gridspec\n",
    "    import seaborn as sns\n",
    "    sns.set_style(\"white\")\n",
    "    sns.set(font_scale=1) ## 修改 seaborn 字体比例\n",
    "    fig = plt.figure(figsize=(6, 6))\n",
    "    k = 2\n",
    "    gs = gridspec.GridSpec(2, k, height_ratios=[1, 1]) ## 多子图构建，2行 k列\n",
    "\n",
    "    ## 调整子图间的整体布局\n",
    "    fig.tight_layout() # 调整整体空白\n",
    "    plt.subplots_adjust(wspace=0.01, hspace=0) # 调整子图间距\n",
    "\n",
    "    ytick=['A', 'C', 'G', 'T']\n",
    "    ## start plot: the left\n",
    "    ## PAM 42-G\n",
    "    pos = 42\n",
    "    data_tuple = get_plot_mismatch1_heatmap_data(pos, data_010, on_data)\n",
    "    data = data_tuple[data_id]\n",
    "    ax0 = plt.subplot(gs[0]) ## 确定子图位置\n",
    "    sns.heatmap(ax=ax0, data=data, mask=mask, vmin=vmin, vmax=vmax, square=True, cbar=False, yticklabels=ytick, cmap=cmap)\n",
    "    ax0.set_xlabel(tf_pos_dict[pos], fontsize=12, weight='bold')\n",
    "    # PAM 43-G\n",
    "    pos = 43\n",
    "    data_tuple = get_plot_mismatch1_heatmap_data(pos, data_010, on_data)\n",
    "    data = data_tuple[data_id]\n",
    "    ax1 = plt.subplot(gs[1]) ## 确定子图位置\n",
    "    sns.heatmap(ax=ax1, data=data, mask=mask, vmin=vmin, vmax=vmax, \n",
    "                square=False, cbar=True, cbar_kws={\"shrink\": 1, 'ticks': [0.05,0.1,0.15, 0.2, 0.25, 0.3]}, \n",
    "                yticklabels=[], cmap=cmap)\n",
    "    ax1.set_xlabel(tf_pos_dict[pos], fontsize=12, weight='bold')\n",
    "    ## title\n",
    "    # plt.title(title, fontsize=10, weight='bold')\n",
    "    mkdir(save_dir)\n",
    "    savefig_path = save_dir + '/%s.%s'%(title, figsuplix)\n",
    "    plt.savefig(savefig_path, dpi=300, bbox_inches = 'tight')\n",
    "\n",
    "\n",
    "## 展示 mismatch 的 relative indel frequence 的分布\n",
    "def plot_multiheatmap(data_010, on_data, cell_line, replicate, suffix, title, save_dir, mask, \n",
    "                      vmax=1, data_id=0, cmap=\"Greys\"):\n",
    "    ## position annotation: transforme\n",
    "    tf_pos_dict = {}\n",
    "    for i in range(21, 44, 1):\n",
    "        if i == 41:\n",
    "            tf_pos_dict[i] = \"N\"\n",
    "        elif (i == 42) | (i == 43):\n",
    "            tf_pos_dict[i] = \"G\"\n",
    "        else:\n",
    "            tf_pos_dict[i] = i-20\n",
    "    \n",
    "    import matplotlib.pyplot as plt\n",
    "    import matplotlib.gridspec as gridspec\n",
    "    import seaborn as sns\n",
    "    sns.set_style(\"white\")\n",
    "    sns.set(font_scale=0.6) ## 修改 seaborn 字体比例\n",
    "    fig = plt.figure(figsize=(15, 1.2))\n",
    "    k = 23\n",
    "    gs = gridspec.GridSpec(2, k, height_ratios=[1, 1]) ## 多子图构建，2行 k列\n",
    "\n",
    "    ## 调整子图间的整体布局\n",
    "    fig.tight_layout() # 调整整体空白\n",
    "    plt.subplots_adjust(wspace=0.01, hspace=0) # 调整子图间距\n",
    "\n",
    "    ytick=['A', 'C', 'G', 'T']\n",
    "    ## start plot: the left\n",
    "    pos = 21\n",
    "    data_tuple = get_plot_mismatch1_heatmap_data(pos, data_010, on_data)\n",
    "    data = data_tuple[data_id]\n",
    "    ax0 = plt.subplot(gs[0]) ## 确定子图位置\n",
    "    sns.heatmap(ax=ax0, data=data, mask=mask, vmax=vmax, square=True, cbar=False, yticklabels=ytick, cmap=cmap)\n",
    "    ax0.set_xlabel(tf_pos_dict[pos], fontsize=12, weight='bold')\n",
    "    for i in range(1, k-1):\n",
    "        pos += 1\n",
    "        data_tuple = get_plot_mismatch1_heatmap_data(pos, data_010, on_data)\n",
    "        data = data_tuple[data_id]\n",
    "        ax1 = plt.subplot(gs[i]) ## 确定子图位置\n",
    "        sns.heatmap(ax=ax1, data=data, mask=mask, vmax=vmax, square=True, cbar=False, yticklabels=[], cmap=cmap)\n",
    "        ax1.set_xlabel(tf_pos_dict[pos], fontsize=10, weight='bold')\n",
    "        if pos == 32:\n",
    "            plt.title(title, fontsize=10, weight='bold')\n",
    "    ## end plot: the right\n",
    "    sns.set(font_scale=0.5) ## 修改 seaborn 字体比例\n",
    "    pos += 1\n",
    "    data_tuple = get_plot_mismatch1_heatmap_data(pos, data_010, on_data)\n",
    "    data = data_tuple[data_id]\n",
    "    ax1 = plt.subplot(gs[k-1]) ## 确定子图位置\n",
    "    sns.heatmap(ax=ax1, data=data, mask=mask, vmax=vmax, square=False, \n",
    "                cbar=True, cbar_kws={\"shrink\": 1, 'ticks': [0.2,0.4, 0.6,0.8]}, \n",
    "                yticklabels=[], cmap=cmap)\n",
    "    ax1.set_xlabel(tf_pos_dict[pos], fontsize=10, weight='bold')\n",
    "    mkdir(save_dir)\n",
    "    savefig_path = save_dir + '/%s-2.%s'%(title, figsuplix)\n",
    "    plt.savefig(savefig_path, dpi=300, bbox_inches = 'tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 展示 mismatch 的 position-substitution 的分布\n",
    "## For position 1 to 21, N, G, G\n",
    "def plot_multiheatmap_ratio_3(data_010, on_data, cell_line, replicate, suffix, title, save_dir, mask, \n",
    "                            vmin=0, vmax=1, data_id=0, cmap=\"Greys\"):\n",
    "    ## position annotation: transforme\n",
    "    tf_pos_dict = {}\n",
    "    for i in range(21, 44, 1):\n",
    "        if i == 41:\n",
    "            tf_pos_dict[i] = \"N\"\n",
    "        elif (i == 42) | (i == 43):\n",
    "            tf_pos_dict[i] = \"G\"\n",
    "        else:\n",
    "            tf_pos_dict[i] = i-20\n",
    "    \n",
    "    import matplotlib.pyplot as plt\n",
    "    import matplotlib.gridspec as gridspec\n",
    "    import seaborn as sns\n",
    "    sns.set_style(\"white\")\n",
    "    sns.set(font_scale=0.6) ## 修改 seaborn 字体比例\n",
    "    fig = plt.figure(figsize=(15, 1.2))\n",
    "    k = 23\n",
    "    gs = gridspec.GridSpec(2, k, height_ratios=[1, 1]) ## 多子图构建，2行 k列\n",
    "\n",
    "    ## 调整子图间的整体布局\n",
    "    fig.tight_layout() # 调整整体空白\n",
    "    plt.subplots_adjust(wspace=0.01, hspace=0) # 调整子图间距\n",
    "\n",
    "    ytick=['A', 'C', 'G', 'T']\n",
    "    ## start plot: the left\n",
    "    pos = 21\n",
    "    data_tuple = get_plot_mismatch1_heatmap_data(pos, data_010, on_data)\n",
    "    data = data_tuple[data_id]\n",
    "    ax0 = plt.subplot(gs[0]) ## 确定子图位置\n",
    "    sns.heatmap(ax=ax0, data=data, mask=mask, vmin=vmin, vmax=vmax, square=True, cbar=False, yticklabels=ytick, cmap=cmap)\n",
    "    ax0.set_xlabel(tf_pos_dict[pos], fontsize=12, weight='bold')\n",
    "    # for i in range(1, k-3):\n",
    "    for i in range(1, k-1):\n",
    "        pos += 1\n",
    "        data_tuple = get_plot_mismatch1_heatmap_data(pos, data_010, on_data)\n",
    "        data = data_tuple[data_id]\n",
    "        ax1 = plt.subplot(gs[i]) ## 确定子图位置\n",
    "        sns.heatmap(ax=ax1, data=data, mask=mask, vmin=vmin, vmax=vmax, square=True, cbar=False, yticklabels=[], cmap=cmap)\n",
    "        ax1.set_xlabel(tf_pos_dict[pos], fontsize=10, weight='bold')\n",
    "        if pos == 32:\n",
    "            plt.title(title, fontsize=10, weight='bold')\n",
    "    ## end plot: the right N\n",
    "    sns.set(font_scale=0.5) ## 修改 seaborn 字体比例\n",
    "    pos += 1\n",
    "    data_tuple = get_plot_mismatch1_heatmap_data(pos, data_010, on_data)\n",
    "    data = data_tuple[data_id]\n",
    "    ax1 = plt.subplot(gs[k-1]) ## 确定子图位置\n",
    "    sns.heatmap(ax=ax1, data=data, mask=mask, vmin=vmin, vmax=vmax, square=False, \n",
    "                cbar=True, cbar_kws={\"shrink\": 1, 'ticks': [0.,0.002,0.004,0.006]}, \n",
    "                yticklabels=[], cmap=cmap)\n",
    "    ax1.set_xlabel(tf_pos_dict[pos], fontsize=10, weight='bold')\n",
    "    mkdir(save_dir)\n",
    "    savefig_path = save_dir + '/%s-total.%s'%(title, figsuplix)\n",
    "    plt.savefig(savefig_path, dpi=300, bbox_inches = 'tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### mismatch 编辑效率分布"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "## step 3: 编辑效率分析 for X0+YA+Z0\n",
    "## 获得 off-target 的相对编辑效率 data\n",
    "## off-target data: fdata\n",
    "## on-target data: odata\n",
    "def mismatch1_off_target_relative_efficiency(fdata, odata, vmin_oe=0.05):\n",
    "    on_dict = {} ## on-target\n",
    "    for index, row in odata.iterrows():\n",
    "        sgRNA_name = row['sgRNA_name']\n",
    "        eff = row['off-target_eff']\n",
    "        on_dict[sgRNA_name] = eff\n",
    "    fdata = fdata[['sgRNA_name', 'all_mutation', 'core_mutation', 'off-target_eff']]\n",
    "    ## 计算 relative eff\n",
    "    fdata['on-target_eff'] = fdata['sgRNA_name'].apply(lambda x: on_dict[x])\n",
    "    fdata1 = fdata.loc[fdata['on-target_eff']>=vmin_oe, :]\n",
    "    fdata1['reltv_eff'] = fdata1.apply(lambda row: row['off-target_eff']/row['on-target_eff'], axis=1)\n",
    "    ## 计算位置\n",
    "    fdata1['position'] = fdata1['core_mutation'].apply(lambda x: mutation_site_range(x)[0])\n",
    "    return fdata1\n",
    "\n",
    "\n",
    "## Get on-target data\n",
    "def get_on_off_target_data(data_010, on_data):\n",
    "    x = 1\n",
    "    data_010_x = data_010.loc[data_010['core_mut_length']==x, :]\n",
    "    #############################################\n",
    "    ## 考虑上下游 mutation\n",
    "    ## on-id\n",
    "    on_data['X'] = on_data['new_mutation'].apply(lambda x: x.split('+')[0])\n",
    "    on_data['Z'] = on_data['new_mutation'].apply(lambda x: x.split('+')[2])\n",
    "    on_data['id'] = on_data.apply(lambda row: row['sgRNA_name'] + '+' + row['X'] + '+' + row['Z'], axis=1)\n",
    "    ## off-id\n",
    "    data_010_x['X'] = data_010_x['new_mutation'].apply(lambda x: x.split('+')[0])\n",
    "    data_010_x['Z'] = data_010_x['new_mutation'].apply(lambda x: x.split('+')[2])\n",
    "    data_010_x['id'] = data_010_x.apply(lambda row: row['sgRNA_name'] + '+' + row['X'] + '+' + row['Z'], axis=1)\n",
    "    \n",
    "    \n",
    "    ## common id \n",
    "    gRNAs = on_data['id'].tolist()\n",
    "    offgRNAs = data_010_x['id'].tolist()\n",
    "    common_gRNAs = list(set(gRNAs) & set(offgRNAs))\n",
    "    print(\"on-target gRNAs number:\", len(set(gRNAs)))\n",
    "    print(\"off-target gRNAs number:\", len(set(offgRNAs)))\n",
    "    print(\"common gRNAs number:\", len(set(common_gRNAs)))\n",
    "    fdata = data_010_x.loc[data_010_x['id'].isin(common_gRNAs), :] ## off-target\n",
    "    odata = on_data.loc[on_data['id'].isin(common_gRNAs), :] ## on-target\n",
    "    print(\"\\n===========================================\")\n",
    "    print(\"off-target.shape:\", fdata.shape)\n",
    "    print(\"on-target.shape:\", odata.shape)\n",
    "    return (fdata, odata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "## plot\n",
    "def plot_relative_eff_with_off_target_pos(fdata1, save_dir):\n",
    "    import matplotlib.pyplot as plt\n",
    "    import seaborn as sns\n",
    "\n",
    "    # 设置默认绘图风格\n",
    "    plt.style.use(\"seaborn-white\")  \n",
    "    fig, ax = plt.subplots(1,1, figsize=(12, 6))\n",
    "\n",
    "    order = [(i) for i in range(21, 44)]\n",
    "    ax = sns.boxplot(x='position', y='reltv_eff', data=fdata1, width=0.4, color='white', \n",
    "                     fliersize=0.5, linewidth=0.5, order=order)\n",
    "    ## 坐标轴不可见\n",
    "    ax.spines['top'].set_visible(False)\n",
    "    ax.spines['right'].set_visible(False)\n",
    "    # # 添加分布散点图 boxplot with jitter\n",
    "    # 通过stripplot添加分布散点图，jitter设置数据间距\n",
    "    ax = sns.stripplot(x='position', y='reltv_eff', data=fdata1, color=\"orange\", jitter=0.1, size=0.8, order=order)\n",
    "    ## 添加水平和垂直直线\n",
    "    plt.axhline(y=1, ls=\"dashed\", c=\"r\", linewidth=1)#添加水平直线\n",
    "    # plt.vlines(19.5, 0, 5, colors = \"r\", linestyles = \"dashed\", linewidth=1)\n",
    "    ## xlabel, ylabel\n",
    "    plt.ylabel(\"Relative editing frequency \\nto on-target efficiency\", fontsize=12, weight='bold')\n",
    "    plt.xlabel(\"Target + PAM Position\", fontsize=12, weight='bold')\n",
    "    plt.ylim(0, 2.3)\n",
    "    ## xticks\n",
    "    xticks = list(range(1, 21, 1)) + ['N', 'G', 'G']\n",
    "    plt.xticks(range(23), xticks, fontsize=10, weight='bold')\n",
    "    title = 'off-target relative editing efficiency of mismatch number 1'\n",
    "    plt.title(title, fontsize=12, weight='bold')\n",
    "    # save_dir = './%s_%s/M4_%s_filtered/pics/%s'%(cell_line, replicate, suffix, 'mismatch')\n",
    "    mkdir(save_dir)\n",
    "    savefig_path = save_dir + '/%s.%s'%(title, figsuplix)\n",
    "    plt.savefig(savefig_path, dpi=300, bbox_inches = 'tight')\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Relative editing efficiency of different mismach number"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 不同 mismatch 长度 off-target 的相对编辑效率\n",
    "def off_target_relative_efficiency_for_mismatch_length(off_data, on_data, on_eff_cutoff=0.05):\n",
    "    import copy\n",
    "    con_data = copy.deepcopy(on_data)\n",
    "    coff_data = copy.deepcopy(off_data)\n",
    "    ## on target data\n",
    "    con_data = con_data.loc[(con_data['all_mutation'] == 'X0+Y0+Z0') & (con_data['off-target_eff']>=on_eff_cutoff), :]\n",
    "    print(\"on_data.shape:\", con_data.shape)\n",
    "    on_dict = {}\n",
    "    for index, row in con_data.iterrows():\n",
    "        on_dict[row['sgRNA_name']] = row['off-target_eff']\n",
    "    ## off target data\n",
    "    coff_data = coff_data.loc[coff_data['sgRNA_name'].isin(list(on_dict.keys())), \n",
    "                           ['sgRNA_name', 'all_mutation', 'core_mut_length', 'off-target_eff']]\n",
    "    coff_data['on-target_eff'] = coff_data['sgRNA_name'] .apply(lambda x: on_dict[x])\n",
    "    coff_data['reltv_eff'] = coff_data.apply(lambda row: row['off-target_eff']/row['on-target_eff'], axis=1)\n",
    "    coff_data.reset_index(drop=True, inplace=True)\n",
    "    return coff_data\n",
    "\n",
    "\n",
    "## 随机选择长度从1 到 8 的 mismach， 计算相对编辑效率\n",
    "def get_random_mismatch_length_data(off_data, count):\n",
    "    import random\n",
    "    stat_data_list = []\n",
    "    for mut_length in range(1, 8):\n",
    "        temp = off_data.loc[off_data['core_mut_length'] == mut_length, :]\n",
    "        random_index = random.sample(list(temp.index), count)\n",
    "        ## DataFrame\n",
    "        stat_data = pd.DataFrame()\n",
    "        stat_data['reltv_eff'] = temp.loc[random_index, :]['reltv_eff'].tolist()\n",
    "        stat_data['mut_length'] = mut_length\n",
    "        stat_data_list.append(stat_data)\n",
    "    stat_data = pd.concat(stat_data_list, axis=0)\n",
    "    return stat_data\n",
    "\n",
    "\n",
    "## plot\n",
    "def plot_mismatch_number_on_reltv_eff(stat_data, save_dir):\n",
    "    import matplotlib.pyplot as plt\n",
    "    import seaborn as sns\n",
    "\n",
    "    # 设置默认绘图风格\n",
    "    plt.style.use(\"seaborn-white\")  \n",
    "    fig, ax = plt.subplots(1,1, figsize=(6, 3))\n",
    "\n",
    "    order = [i for i in range(1, 8)]\n",
    "    ax = sns.boxplot(x='mut_length', y='reltv_eff', data=stat_data, width=0.4, color='white', \n",
    "                     fliersize=0.5, linewidth=0.5, order=order)\n",
    "    ## 坐标轴不可见\n",
    "    ax.spines['top'].set_visible(False)\n",
    "    ax.spines['right'].set_visible(False)\n",
    "    ## 添加分布散点图 boxplot with jitter\n",
    "    ## 通过stripplot添加分布散点图，jitter设置数据间距\n",
    "    ax = sns.stripplot(x='mut_length', y='reltv_eff', data=stat_data, color=\"orange\", jitter=0.1, size=0.8, order=order)\n",
    "    plt.ylabel(\"Relative editing frequency \\nto on-target efficiency\", fontsize=12, weight='bold')\n",
    "    plt.xlabel(\"Number of mismatches\", fontsize=12, weight='bold')\n",
    "    plt.ylim(-0.08, 1.58)\n",
    "    ## xticks\n",
    "    xticks = order\n",
    "    plt.xticks(range(len(xticks)), xticks, fontsize=10, weight='bold')\n",
    "    title = 'The effect of mismatch number in Target PAM Region on editing efficiency'\n",
    "    plt.title(title, fontsize=12, weight='bold')\n",
    "    mkdir(save_dir)\n",
    "    savefig_path = save_dir + '/%s.%s'%(title, figsuplix)\n",
    "    stat_data.to_csv(save_dir + \"/%s.csv\"%(title), index=False)\n",
    "    s = stat_data.loc[(stat_data['mut_length']==1), :].shape[0]\n",
    "    a = stat_data.loc[(stat_data['reltv_eff']>=0.5) & (stat_data['mut_length']==1), :].shape[0]\n",
    "    print(a/s)\n",
    "    plt.savefig(savefig_path, dpi=300, bbox_inches = 'tight')\n",
    "    plt.show()\n",
    "########################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "## load data\n",
    "def Obtain_mismatch_data(read_data_dir, mut_type):\n",
    "    if cell_line == 'H1':\n",
    "        reads_num, barcode_num = 400, 10\n",
    "    else:\n",
    "        reads_num, barcode_num = 200, 10\n",
    "    ## read\n",
    "    data_path = '%s/off-target.%s.accurate.offseq'%(read_data_dir, mut_type)\n",
    "    data = pd.read_csv(data_path, sep='\\t')\n",
    "    data = data.loc[(data['reads_num']>=reads_num) & (data['barcode_num']>=barcode_num), :]\n",
    "    data.reset_index(drop=True, inplace=True)\n",
    "    print('columns:', data.columns.tolist())\n",
    "    print(data.shape)\n",
    "\n",
    "    ## data 根据 up, core, down 区域突变与否划分数据集 \n",
    "    data_000, data_100, data_010, data_001, data_110, data_101, data_011, data_111, count_dict = mutation_region_splitting(data)\n",
    "    return (data, data_010, reads_num, barcode_num)\n",
    "\n",
    "\n",
    "## obtain on target data\n",
    "def obtaim_on_target_data(read_data_dir, reads_num, barcode_num):\n",
    "    on_data_path = '%s/off-target.perfect.accurate.offseq'%(read_data_dir)\n",
    "    on_data = pd.read_csv(on_data_path, sep='\\t')\n",
    "    on_data = on_data.loc[(on_data['all_mutation']=='X0+Y0+Z0') & (on_data['off-target_eff']>=0.05), :]\n",
    "    on_data = on_data.loc[(on_data['reads_num']>=reads_num) & (on_data['barcode_num']>=barcode_num), :]\n",
    "    on_data.reset_index(drop=True, inplace=True)\n",
    "    return on_data\n",
    "\n",
    "\n",
    "## summary_plot\n",
    "def plot_all_mismatch(cell_line, replicate, data, data_010, on_data):\n",
    "    import copy \n",
    "    print(\"\\n## plot distribution of off-target mismatch number\")\n",
    "    core_mut_length_dict = get_off_target_mismatch_number(data_010)\n",
    "    save_dir = './%s/off-target/FigSupp-distribution of off-target mismatch num/%s-%s'%(figsuplix, cell_line, replicate)\n",
    "    mkdir(save_dir)\n",
    "    plot_distribution_off_target_mismacth_num(core_mut_length_dict, save_dir)\n",
    "    ##################################################\n",
    "    \n",
    "    ## plot distribution of off-target mismatch among Target + PAM regions\n",
    "    print(\"\\n## plot distribution of off-target mismatch among Target + PAM regions\")\n",
    "    save_dir = './%s/off-target/FigSupp-distribution of mismatch with position/%s-%s'%(figsuplix, cell_line, replicate)\n",
    "    mkdir(save_dir)\n",
    "    plot_distribution_off_target_mismatch_with_positions(data, save_dir)\n",
    "    ##################################################\n",
    "    \n",
    "    ## plot distribution of substitution types\n",
    "    print(\"\\n## plot distribution of substitution types\")\n",
    "    save_dir = './%s/off-target/FigSupp-distribution of substitution types/%s-%s'%(figsuplix, cell_line, replicate)\n",
    "    mkdir(save_dir)\n",
    "    plot_distribution_substitution_types(data, save_dir)\n",
    "    ##################################################\n",
    "    \n",
    "    ## data preparation for relative off-target efficiency\n",
    "    print(\"\\n## 对 mismatch position & substitution 的分布分析 \")\n",
    "    on_data_copy = copy.deepcopy(on_data)\n",
    "    pos = 21\n",
    "    # for pos in range(21, 44):\n",
    "    print('\\nposition:', pos)\n",
    "    print(\"===============================\")\n",
    "    plot_data, stat_count = get_plot_mismatch1_heatmap_data(pos, data_010, on_data_copy)\n",
    "    mask = help_creak_mask(plot_data)\n",
    "    ## 对 mismatch position & substitution 的分布分析 \n",
    "    ## type distribution\n",
    "    save_dir = './%s/off-target/Fig20-heatmap of distribution of substitution/%s-%s'%(figsuplix, cell_line, replicate)\n",
    "    mkdir(save_dir)\n",
    "    title0 = 'Ratio distribution of substitution of off-target mismatch-1'\n",
    "    plot_multiheatmap_ratio_3(data_010, on_data, cell_line, replicate, suffix, title0, save_dir, mask, \n",
    "                          vmin=0, vmax=0.008, data_id=1, cmap=\"Greys\")\n",
    "    ## relative efficiency\n",
    "    ## 对 mismatch position & substitution 的分布分析\n",
    "    title1 = 'off-target relative efficiency of substitution of mismatch-1'\n",
    "    save_dir = './%s/off-target/Fig21-heatmap of distribution of relative efficiency/%s-%s'%(figsuplix, cell_line, replicate)\n",
    "    mkdir(save_dir)\n",
    "    plot_multiheatmap(data_010, on_data_copy, cell_line, replicate, suffix, title1, save_dir, mask, \n",
    "                      vmax=1, data_id=0, cmap=\"Greys\")\n",
    "    ##################################################\n",
    "    \n",
    "    ## 获得 off-target 的相对编辑效率 data\n",
    "    print(\"\\n## plot_relative_eff_with_off_target_pos\")\n",
    "    on_data.rename(columns={\"on-target_eff\": \"off-target_eff\"}, inplace=True)\n",
    "    on_data_copy = copy.deepcopy(on_data)\n",
    "    fdata, odata = get_on_off_target_data(data_010, on_data_copy)\n",
    "    fdata1 = mismatch1_off_target_relative_efficiency(fdata, odata, vmin_oe=0.05)\n",
    "    ## plot_relative_eff_with_off_target_pos\n",
    "    save_dir = './%s/off-target/Fig21-heatmap of distribution of relative efficiency/%s-%s'%(figsuplix, cell_line, replicate)\n",
    "    mkdir(save_dir)\n",
    "    plot_relative_eff_with_off_target_pos(fdata1, save_dir)\n",
    "    ##################################################\n",
    "    \n",
    "    ## 不同 mismatch 长度 off-target 的相对编辑效率\n",
    "    print(\"\\n## plot Relative editing efficiency of different mismach number\")\n",
    "    on_data_copy = copy.deepcopy(on_data)\n",
    "    off_data = off_target_relative_efficiency_for_mismatch_length(data_010, on_data_copy, on_eff_cutoff=0.05)\n",
    "    print(off_data.shape)\n",
    "    ## 随机选择 count 个长度从1 到 8 的 mismach， 计算相对编辑效率\n",
    "    count = 500\n",
    "    stat_data = get_random_mismatch_length_data(off_data, count)\n",
    "    print(stat_data.shape)\n",
    "    save_dir = './%s/off-target/Fig22-effect of mismatch number/%s-%s'%(figsuplix, cell_line, replicate)\n",
    "    mkdir(save_dir)\n",
    "    plot_mismatch_number_on_reltv_eff(stat_data, save_dir)\n",
    "    return off_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
